File: connectors/peertube-ingest/src/batch.ts
/**
 * Script batch, execute en cron (etude 2.12 ligne 589 : "ce depot peut se faire en tache
 * periodique -- batch quotidien -- plus simple a operer, sans risque meme en cas de
 * panne temporaire du connecteur"). Liste les objets deposes depuis la derniere execution
 * et les depose vers PeerTube.
 *
 * Usage : `node dist/batch.js [--since=2026-09-04T00:00:00Z]`
 * Sans `--since`, retombe sur "les dernieres 25 heures" (marge d'1h pour couvrir un
 * decalage d'execution du cron quotidien sans trou de couverture).
 */
import { filterVideoObjects, ingestAll } from "./ingest";
import { getObjectStream, getObjectTags, listRecentObjects } from "./minio-client";
import { uploadToPeerTube } from "./peertube-client";

function resolveSince(argv: string[]): string {
  const arg = argv.find((a) => a.startsWith("--since="));
  if (arg) {
    return arg.slice("--since=".length);
  }
  return new Date(Date.now() - 25 * 60 * 60 * 1000).toISOString();
}

async function main(): Promise<void> {
  const since = resolveSince(process.argv.slice(2));
  // eslint-disable-next-line no-console
  console.log(`[peertube-ingest batch] scanning objects modified since ${since}`);

  const objects = await listRecentObjects(since);
  const candidates = filterVideoObjects(objects);
  // eslint-disable-next-line no-console
  console.log(`[peertube-ingest batch] ${candidates.length} candidate(s) to ingest`);

  const results = await ingestAll(candidates, { getObjectTags, getObjectStream, uploadToPeerTube });

  const failed = results.filter((r) => !r.uploaded);
  // eslint-disable-next-line no-console
  console.log(
    `[peertube-ingest batch] done: ${results.length - failed.length} uploaded, ${failed.length} failed`
  );
  if (failed.length > 0) {
    // eslint-disable-next-line no-console
    console.error(JSON.stringify(failed, null, 2));
    process.exitCode = 1;
  }
}

if (require.main === module) {
  main().catch((err) => {
    // eslint-disable-next-line no-console
    console.error("[peertube-ingest batch] fatal error", err);
    process.exitCode = 1;
  });
}

export { resolveSince };

File: connectors/peertube-ingest/app/batch.py
"""Batch script, run via cron (study 2.12 line 589: "this upload can be done as a
periodic task -- daily batch -- simpler to operate, with no risk even if the
connector has a temporary outage"). Lists objects uploaded since the last run and
uploads them to PeerTube.

Usage: `python -m app.batch [--since=2026-09-04T00:00:00Z]`
Without `--since`, falls back to "the last 25 hours" (1h margin to cover a
scheduling drift of the daily cron without a coverage gap).
"""

from __future__ import annotations

import asyncio
import json
import sys
from dataclasses import asdict
from datetime import datetime, timedelta, timezone
from typing import Any, Dict, List, Sequence

import httpx

from .ingest import IngestDeps, filter_video_objects, ingest_all
from .s3_client import get_object_stream, get_object_tags, list_recent_objects
from .peertube_client import upload_to_peertube


def resolve_since(argv: Sequence[str]) -> str:
    for arg in argv:
        if arg.startswith("--since="):
            return arg[len("--since=") :]
    return (datetime.now(timezone.utc) - timedelta(hours=25)).isoformat()


async def main(argv: Sequence[str] | None = None) -> int:
    argv = list(argv if argv is not None else sys.argv[1:])
    since = resolve_since(argv)
    print(f"[peertube-ingest batch] scanning objects modified since {since}")

    objects = await list_recent_objects(since)
    candidates = filter_video_objects(objects)
    print(f"[peertube-ingest batch] {len(candidates)} candidate(s) to ingest")

    async with httpx.AsyncClient() as http_client:

        async def _upload(**kwargs: Any) -> Dict[str, str]:
            return await upload_to_peertube(http_client=http_client, **kwargs)

        deps = IngestDeps(
            get_object_tags=get_object_tags,
            get_object_stream=get_object_stream,
            upload_to_peertube=_upload,
        )
        results = await ingest_all(candidates, deps)

    failed = [r for r in results if not r.uploaded]
    print(
        f"[peertube-ingest batch] done: {len(results) - len(failed)} uploaded, "
        f"{len(failed)} failed"
    )
    if failed:
        print(json.dumps([asdict(r) for r in failed], indent=2), file=sys.stderr)
        return 1
    return 0


def run() -> None:
    try:
        exit_code = asyncio.run(main())
    except Exception as err:  # noqa: BLE001 - mirrors the TS top-level catch
        print(f"[peertube-ingest batch] fatal error: {err}", file=sys.stderr)
        exit_code = 1
    sys.exit(exit_code)


if __name__ == "__main__":
    run()
